### DISCUSSION TITLE
Delete empty lines after deleting brackets

### DISCUSSION THREAD

**Discussion by xeluxee** [kylechui/nvim-surround#163] [Q&A]
I've previously used vim-sandwich, that can be configured to delete empty lines after deleting brackets. Take the following example:
``` c
if (x)
{
	y = z; *cursor on this line*
}
//
```
After pressing `ds}` with cursor on line 3:
``` c
if (x)

	y = z;

//
```
But I'd like lines 2 and 4 to be deleted:
``` c
if (x)
	y = z;
//
```
Is there a way to delete empty lines after deleting brackets?

**Comment by kylechui** [comment]
The way to do this would be to check if a whitespace character exists (or newline character specifically, if that's what you want) in the `delete` key, as follows:

```lua
require("nvim-surround").setup({
    surrounds = {
        ["{"] = {
            delete = "^(.%s*)().-(%s*.)()$",
        },
    }
}
```

This can be repeated for the closing curly brace as well.

**Comment by xeluxee** [comment]
Thanks for the suggestion, but it isn't working in the above example: when I press `ds{` nothing happens.
Furthermore I've noticed that it deletes the text between two opening brackets (`{text{`):

``` c
{
	if (x)
	{
		y = z; *cursor on this line*
	}
}
```
Press `ds{`:
``` c
if (x)

		y = z;
	}
}
```

Anyway what's the string being matched by this lua pattern? Does it include newlines, so that they can be deleted?

**Reply by kylechui** [reply]
The string being matched should say "delete the pair of curly braces, as well as one white space character between them, if it exists". It seems to be doing what is intended on my end. If you're using a tagged version of the plugin, probably try following `main` instead. I suspect it has to do with a configuration option I toggled *a while* ago.

**Reply by xeluxee** [reply]
Yeah, I was using a tagged version.
Now the first empty line is deleted, but not the second one.

> The string being matched should say "delete the pair of curly braces, as well as one white space character between them, if it exists".

Sorry, I wasn't asking its meaning. Instead it's unclear to me what portion of buffer text is considered to be matched with this lua pattern

**Reply by kylechui** [reply]
The delete/change actions are broken up into two steps as follows:

* Find the "parent selection" that includes the left/right surrounding pair
  * For `{`, the default is to use the `a{` textobject
  * In your example, it should select the inner pair of curly braces, and their internal contents
* Isolate the left/right surrounding pair from the provided "parent selection"
  * For Lua patterns, the parent selection's contents are joined with `\n` as a separator
  * In your case, the string would be something like `{\n      y = z;\n      }` (incorrect number of spaces but you get the point
  * Then the first and third match groups are used to actually determine what to delete

I realized that I made a mistake in the Lua pattern, both `?` symbols should be `*` instead. In general, you should be able to play around with the Lua pattern until you get the behavior that you want; I can't guarantee my pattern will be correct for all use cases based on just a handful of examples. I've updated my original response with the correct pattern, if you want to try that out.

**Reply by xeluxee** [reply]
Thanks, now it's clear
> I realized that I made a mistake in the Lua pattern, both `?` symbols should be `*` instead.

It works better. But still there are some edge cases where it fails.
For example when the opening bracket isn't on a new line, the first line inside the block is joined with the line above, because the newline soon after the opening bracket is deleted.
``` c
if (x) {
	y = z;
}
```
After pressing `ds{` with cursor on line 2:
``` c
if (x) y = z;
```
Lua patterns don't seem suitable for this task, furthermore I'd need to copy that pattern for many different surrounds in the setup function.

Why not using the function `indent_lines` to remove empty lines and to post-process text in general?
I've played around with it to come out with the following solution:
``` lua
indent_lines = function(start, stop)
  local b = vim.bo
  -- Only re-indent the selection if a formatter is set up already
  if start <= stop and (b.equalprg ~= "" or b.indentexpr ~= "" or b.cindent or b.smartindent or b.lisp) then
    vim.cmd(string.format("silent normal! %dG=%dG", start, stop))
  end

  if start < stop then
    local function clear_line(line) -- strip trailing whitespaces and delete line if empty
      local l, c = string.gsub(vim.api.nvim_buf_get_lines(0, line - 1, line, false)[1], "%s+$", "")
      if l == "" then
        vim.api.nvim_buf_set_lines(0, line - 1, line, false, {})
      elseif c ~= 0 then
        vim.api.nvim_buf_set_lines(0, line - 1, line, false, { l })
      end
    end
    clear_line(stop)
    clear_line(start)
  end
end
```
The function `clear_line` also deletes trailing whitespaces, so in the above example there'll be no space after `if (x)` when the opening bracket gets deleted.

**Reply by kylechui** [reply]
Oh, I see now. You want to remove all whitespace-only lines from the selection after deleting the surrounding pair, correct? I'm averse to using `indent_lines` to performing this task, as this "feature" that you're proposing is not any more "correct" than leaving the whitespace lines where they are. Hence I would say that this should be done from a user configuration standpoint, rather than by modifying the core functionality of the plugin (without maintaining existing behavior). However, a global setup option (say `remove_whitespace_lines`) seems like it could work, where we first delete the delimiter pair according to the current heuristic (just the pair, leave everything else alone), then come back and use the global flag to determine whether or not to remove the "empty" lines (if any exist). However, I'm a bit busy at the moment, and I would rather prioritize working on Treesitter/Query-related features before I put more time/energy into thinking about this problem.

**Reply by xeluxee** [reply]
> Oh, I see now. You want to remove all whitespace-only lines from the selection after deleting the surrounding pair, correct?

I don't want to remove **all** whitespaces-only lines, but **only** the starting and ending line if they become whitespaces-only after deleting brackets.

> However, a global setup option (say `remove_whitespace_lines`) seems like it could work

Nice idea. Wouldn't `remove_empty_lines` be more meaningful?

**Reply by kylechui** [reply]
> I don't want to remove all whitespaces-only lines, but only the starting and ending line if they become whitespaces-only after deleting brackets.

Yeah that's what I meant to say, sorry for the confusion.

> Nice idea. Wouldn't remove_empty_lines be more meaningful?

I'm not 100% sure on naming since empty (to me) implies literally an empty line, without even whitespace. But naming's another problem that I'll think about if/when I get around to working on this.
